src/warc2zim/utils.py

```diff
@@ -125,55 +130,77 @@ def get_record_encoding(record: ArcWarcRecord) -> str | None:
         return m.group("encoding")
 
 
-def to_string(input_: str | bytes, encoding: str | None) -> str:
+def to_string(input_: str | bytes, encoding: str | None) -> StringConversionResult:
     """
     Decode content to string, trying to be the more tolerant possible to invalid
     declared encoding.
 
-    This try decode the content using 3 methods:
+    Returns the decoded content and the encoding used.
+
+    This try decode the content using 3 methods, failing on first error:
      - From http headers in the warc record (given as `encoding` argument)
      - From encoding declaration inside the content (hopping that content can be
        losely decode using ascii to something usable)
      - From statistical analysis of the content (made by chardet)
 
+    If all these methods fails, try again with the encoding passed via http headers but
+    ignore all unrecognized characters.
+
     """
-    tried_encodings = set()
+    http_encoding = encoding
+
+    tried_encodings: set[str] = set()
     if isinstance(input_, str):
-        return input_
+        return StringConversionResult(input_, None, False)
 
     if not input_:
         # Empty bytes are easy to decode
-        return ""
+        return StringConversionResult("", None, False)
 
     if encoding:
         try:
-            return input_.decode(encoding)
+            return StringConversionResult(input_.decode(encoding), encoding, False)
         except (ValueError, LookupError):
             tried_encodings.add(encoding)
             pass
 
-    # Detect encoding from content.
+    # Search for encoding from content first bytes based on regexp
     content_start = input_[:1024].decode("ascii", errors="replace")
     if m := ENCODING_RE.search(content_start):
         encoding = m.group("encoding")
         if encoding and encoding not in tried_encodings:
             try:
-                return input_.decode(encoding)
+                return StringConversionResult(input_.decode(encoding), encoding, False)
             except (ValueError, LookupError):
                 tried_encodings.add(encoding)
                 pass
 
-    encodings = (
-        encoding
-        for e in chardet.detect_all(input_)
-        if (encoding := e["encoding"]) and encoding not in tried_encodings
-    )
+    # Try to detect the most probable encoding with chardet (and only most probable
+    # one, since otherwise we will likely find an encoding which pass but produces only
+    # garbage with most characters badly decoded just due to a wrongly encoded character
+    # see https://github.com/openzim/warc2zim/issues/221)
+    chardet_encodings = chardet.detect_all(input_)
+    if len(chardet_encodings) > 0:
+        chardet_encoding = chardet_encodings[0]["encoding"]
```

Comment:
Because chardet inner working is different when we request all encodings than when we request only one. The most probable encoding when we request all is not the same as the one returned when we request only one encoding. Don't ask me why.

Reply:
Interesting ; can you then add a comment mentioning it?

Reply:
Sure

Reply:
Btw, it is already trapped in the test cases (this is how I realized it indeed)